### ISSUE TITLE
Parmeter Tuning based on out of sample error

### ISSUE THREAD

**Issue by sreepunith** [mlr-org/mlr#2019]
This maybe a very naïve question. I was wondering if there is a way to tune parameter based on the error on the hold out sample rather than out of fold cross validated error. 
I was thinking if I can create a task based on the holdout data, which is not included in the training task. Can I estimate the performance of the tuned parameter on those ?

Thanks

**Comment by jakob-r** [comment]
That would be something like this?

```r
library(mlr)
rd = makeResampleDesc("Holdout")
ctrl = makeTuneControlGrid(same.resampling.instance = TRUE, resolution = 4)
tune.res = tuneParams(learner = "classif.randomForest", task = iris.task, resampling = rd, par.set = makeNumericParamSet("mtry", 1, 1, 4), control = ctrl)
```

**Comment by berndbischl** [comment]
what @jakob-r said.

you can specify very different ways to create repeated data splits via all the resampling strategies. 

a few notes: 
- you have to include the holdout data in the task for this (comment if you think this would be a problem, beacuse i dont see why)
- in Jakobs's example code, the holdout set is chosen randomly. I guess you might dislike this. use 
makeFixedHoldoutInstance instead, then.  and just define the indices in your task that are the test set.

hope this helps. 

reopen if this does not answer your problem!

**Comment by sreepunith** [comment]
Thank you for those suggestions. I was thinking if there was a way in which, I could do the tuning based on cross validation and then, try the tuned parameters on a holdout set. I want to use the score from the out od sample test to be minimized. I have looked at nested resampling, but it just gives a measure of how well the model does.

**Comment by berndbischl** [comment]
hmm. now i am getting confused, what you are asking.
i mean you can EITHER tune the params based on a CV or on a holdout set. we have shown you how you can do a) or b). and of course, if you tune based on a CV, you can validate the params on a freh test set. 

but in your last question, it sounds like you want to first tune with a CV and then tune "again" on a holdout set, and i am anfraid i dont understand what that means.

can you write down more precisely what you want, maybe in a step-by-step basis?

**Comment by sreepunith** [comment]
OK. Let me try it be more clearer:
Step1: Split data into training and validation
Step2: Create Task using training data
Step3: Using the task tune parameters by CV
Step4: For each iteration, calculate the error(mae) for the tuned parameter using the validation data

I know if I do tuning via CV, the model gives me error based on the out of fold sample, I want the model to give me the error on the validation data, which is not used in tuning. 

I hope I didn't confuse you any further.

Thank you

**Comment by jakob-r** [comment]
Maybe you should have a look at https://github.com/jakob-r/mlrOverfit
However, be aware that these results should be interpreted with care.

**Comment by sreepunith** [comment]
Thank you. Can you tell me how can I validate on a fresh dataset, if I tune via CV? Do I need to create a new task or can I subset the data from the task?

**Comment by jakob-r** [comment]
For mlrOverfit you would have to combine these datasets first and make a fixed holdout that marks the "fresh dataset" and pass it to `resampleOverfit()`.

**Comment by sreepunith** [comment]
Thank you and if I had to do it in mlr? How would I do it?

**Comment by jakob-r** [comment]
Exactly how it is done in mlrOverfit. Take a look in the source code of mlrOverfit to find out how it's done.

**Comment by berndbischl** [comment]
> Step4: For each iteration, calculate the error(mae) for the tuned parameter using the validation data

do you mean here: for each iteration of the *optimization*, also additionally evaluate the param vector on the validation set? to track what happens there?

**Comment by sreepunith** [comment]
Yes!

**Comment by sreepunith** [comment]
I looked at mlrOverfit, unfortunately it doesnt work for me. I wanted to use makefixedHoldoutInstance instead of standard holdout. With that I cant have a CV resampling scheme in the outler layer.

**Comment by berndbischl** [comment]
Hi.

1) What you want to do is not "perfectly" supported by mlr at the moment. What you can do is track the error during tuning, thats being logged into the opt.path of the tune.result. 
And you also get the result of the *best* tuned parameters, evaled on a fresh testset. 
You can do that manually (tuneParams, then train, then predict) or with "nested resampling", with a TuneWrapper. All of that is in the tutoriall.

2) But you want to track performance for all evaled params on a different set. 

3) Here is some code to show, how it CAN  be done with mlr - although I am not too proud of the structure of the code:

```
# quickly split data
idx1 = seq(1, 150, 2)
idx2 = seq(2, 150, 2)
d1 = iris[idx1,]
d2 = iris[idx2,]
# task on indices idx1
task = makeClassifTask(data = d1, target = "Species")

# stupid param set
lrn = makeLearner("classif.ksvm")
ps = makeParamSet(
  makeNumericParam("C", lower = -10, upper = 10, trafo = function(x) 2^x)
)
# stupid tuning
ctrl = makeTuneControlRandom(maxit = 10)


########### HERE THE MAGIC STARTS ##################
m1 = mmce
# we create 2nd dummy measure to login to
m2 = mmce; m2$id = "extra"
ms = list(m1, m2)

# we now overwrite the resample function in  tuning
myresample = function(learner, task, resampling, measures, show.info) {
  # we generate the normal resample call
  r = resample(learner, task, resampling, measures, show.info = show.info)
  # and now we train on the complete task, predict extra holdout
  mod = train(learner, task)
  pred = predict(mod, newdata = d2)
  # and overwrite our dummy measure in the optpath
  p = performance(pred, measures = mmce)
  r$aggr[2] = p
  return(r)
}


# normal tuneParams call, but with overwritten resample.fun
tr = tuneParams(lrn, task, cv2, par.set = ps, control = ctrl, measures = ms,
  resample.fun = myresample)

opdf = as.data.frame(tr$op)
print(opdf)
```

```
Loading mlr
Loading required package: ParamHelpers
          C mmce.test.mean extra.test.mean dob eol error.message exec.time
1   6.94906        0.08001         0.06667   1  NA          <NA>     0.976
2  -0.04945        0.09317         0.05333   2  NA          <NA>     0.091
3  -2.24182        0.18670         0.04000   3  NA          <NA>     0.029
4  -5.07102        0.70661         0.10667   4  NA          <NA>     0.032
5  -7.77807        0.70661         0.10667   5  NA          <NA>     0.028
6  -2.20011        0.17354         0.05333   6  NA          <NA>     0.029
7   1.43871        0.07966         0.05333   7  NA          <NA>     0.032
8  -5.66214        0.70661         0.10667   8  NA          <NA>     0.029
9  -1.10464        0.10704         0.05333   9  NA          <NA>     0.029
10 -5.64019        0.70661         0.10667  10  NA          <NA>     0.032

```

**Comment by berndbischl** [comment]
4) maybe i have a better idea to achieve this with less terrible code. let me think quickly

**Comment by jakob-r** [comment]
This sentence
> With that I cant have a CV resampling scheme in the outler layer.

contradicts with what you mentioned here

>Step1: Split data into training and validation
>Step2: Create Task using training data
>Step3: Using the task tune parameters by CV
>Step4: For each iteration, calculate the error(mae) for the tuned parameter using the validation data

**Comment by berndbischl** [comment]
Here is a much better solution

```
# lets shuffle iris a bit
d = iris[sample(150),]
# take ALL data in task, including extra holdout
task = makeClassifTask(data = d, target = "Species")

# stupid param set
lrn = makeLearner("classif.ksvm")
ps = makeParamSet(
  makeNumericParam("C", lower = -10, upper = 10, trafo = function(x) 2^x)
)
# stupid tuning
ctrl = makeTuneControlRandom(maxit = 10)


########### HERE THE MAGIC STARTS ##################

# create custom data splitting
# 1) we train on first "third", eval on 2d third (for this we tune)
# 2) we train on first "third", eval on 3rd third (extra holdout eval to check)
rin = makeResampleInstance("Subsample", iters= 2, task = task)
rin$train.inds[[1]] = 1:50
rin$test.inds[[2]] = 51:100
rin$train.inds[[1]] = 1:50
rin$test.inds[[2]] = 101:150

# we now create 2 different "aggregations"
# for our measure of choice (here: mmce)
# m1 picks out the perf of the 1st train/test split
# m2 picks out the perf of the 2nd train/test split
m1 = m2 = mmce
m1 = setAggregation(m1, makeAggregation(id = "split1", properties = "req.test",
  fun = function(task, perf.test, perf.train, measure, group, pred) perf.test[1]
))
m2 = setAggregation(m2, makeAggregation(id = "split2", properties = "req.test",
  fun = function(task, perf.test, perf.train, measure, group, pred) perf.test[2]
))
ms = list(m1, m2)

# tuneParams will optimize m1 (but log all other measures, so here m2, too!)
tr = tuneParams(lrn, task, cv2, par.set = ps, control = ctrl, measures = ms)

opdf = as.data.frame(tr$op)
print(opdf)
```

**Comment by berndbischl** [comment]
that should solve your poblem, HTH!

**Comment by jakob-r** [comment]
@berndbischl Your example looks wrong. The indices for `rin` for instance. Don't you have to replace `cv2` with `rin` in your example? 

Also: Won't the model be trained twice on the training set (1:50) because of the two iterations?

**Comment by berndbischl** [comment]
> @berndbischl Your example looks wrong. The indices for rin for instance. Don't you have to replace cv2 with rin in your example?

thx fixed. and found more typos myself.

> Also: Won't the model be trained twice on the training set (1:50) because of the two iterations?

thats true.

1) either live with this, or use my first version above. 

2) in some cases, the train.set for the 2nd eval would be different. like 
rin$train.inds[[2]] = 1:100
thats more realistic and closer to what happens in nested resampling where at the end you train on these 2 sets, so maybe you want to extactly that for all param configs.
then the 2nd trainig step is unavoidable. 

3) what mlr really should provide is a "plugin function" that is called after every tuning step, so you can log whatever you want into the optpath. sadly we dont have that yet.

**Comment by berndbischl** [comment]
and here now is a hopfully final version :)
this only trains once.

```
load_all()

d = iris[sample(150),]
task = makeClassifTask(data = d, target = "Species")
lrn = makeLearner("classif.ksvm")
ps = makeParamSet(
  makeNumericParam("C", lower = -10, upper = 10, trafo = function(x) 2^x)
)
ctrl = makeTuneControlRandom(maxit = 10)

# the last 2 thirds should contain 2 eval sets
rin = makeFixedHoldoutInstance(train.inds = 1:50, test.inds = 51:150, size = 150)

# now lets measure on eval set 1
m1 = makeCustomResampledMeasure("split1", aggregation.id = "",
  properties = c("classif", "classif.multi"),
  fun = function(task, group, pred, extra.args) {
    p = pred; p$data = p$data[1:50,]
    performance(p, mmce)
  }
)
# and eval set 2
m2 = makeCustomResampledMeasure("split2", aggregation.id = "",
  properties = c("classif", "classif.multi"),
  fun = function(task, group, pred, extra.args) {
    p = pred; p$data = p$data[51:100,]
    performance(p, mmce)
  }
)
ms = list(m1, m2)
tr = tuneParams(lrn, task, rin, par.set = ps, control = ctrl, measures = ms)
opdf = as.data.frame(tr$op)
print(opdf)

```

**Comment by sreepunith** [comment]
Thank you for all your help. I really appreciate it. In the last version of the code: for eval set 2, why are the indices of eval set from 51:100, p$data[51:100,], instead of 51:150 ? If I try 51:150, I get Error in if (any(co)) { : missing value where TRUE/FALSE needed.

Thanks again

**Comment by sreepunith** [comment]
Something like this

load_all()
data(BostonHousing, package = "mlbench")
data<-subset(BostonHousing, select=-chas)
task = makeRegrTask(data = data, target = "medv")
lrn = makeLearner("regr.ksvm")
ps = makeParamSet(
  makeNumericParam("C", lower = -10, upper = 10, trafo = function(x) 2^x)
)
ctrl = makeTuneControlRandom(maxit = 10)

# the last 2 thirds should contain 2 eval sets
rin = makeFixedHoldoutInstance(train.inds = 1:400, test.inds = 401:506, size = 506)

# now lets measure on eval set 1
m1 = makeCustomResampledMeasure("split1", aggregation.id = "",
                                properties = c("regr"),
                                fun = function(task, group, pred, extra.args) {
                                  p = pred; p$data = p$data[1:400,]
                                  performance(p, mae)
                                }
)
# and eval set 2
m2 = makeCustomResampledMeasure("split2", aggregation.id = "",
                                properties = c("regr"),
                                fun = function(task, group, pred, extra.args) {
                                  p = pred; p$data = p$data[401:506,]
                                  performance(p, mae)
                                }
)
ms = list(m1, m2)
tr = tuneParams(lrn, task, rin, par.set = ps, control = ctrl, measures = ms)
opdf = as.data.frame(tr$op)
print(opdf)

**Comment by sreepunith** [comment]
@jakob-r 
I tried mlrOverfit, but I wanted fixedholdoutInstance instead of holdout: If I did that I get the following error
 Size of data set: 71312 and resampling instance: 89140 differ!

**Comment by jakob-r** [comment]
Can you open an issue there with an example?

**Comment by berndbischl** [comment]
> n the last version of the code: for eval set 2, why are the indices of eval set from 51:100, p$data[51:100,], instead of 51:150 ? If I try 51:150, I get Error in if (any(train)) { : missing value where TRUE/FALSE needed.

because the test test, that you predict on, has 100  elements?

**Comment by sreepunith** [comment]
When I try to run tuneparams with custom resample function in parallel, I get an error
00001: Error in resample(learner, task, resampling, measures, show.info = show.info) : 
  could not find function "resample". 
I tried parallelExport('myresample',level='mlr.resample'), it doesnt work either
